src/lib/message/L1Transaction.ts

```diff
@@ -193,15 +242,31 @@ export class L1TransactionReceipt implements TransactionReceipt {
    */
   public async getL1ToL2Messages<T extends SignerOrProvider>(
     l2SignerOrProvider: T
-  ): Promise<L1ToL2MessageReaderOrWriter<T>[]>
+  ): Promise<L1ToL2MessageReaderOrWriter<T>[] | ClassicL1ToL2MessageReader[]>
   public async getL1ToL2Messages<T extends SignerOrProvider>(
     l2SignerOrProvider: T
-  ): Promise<L1ToL2MessageReader[] | L1ToL2MessageWriter[]> {
+  ): Promise<
+    L1ToL2MessageReader[] | L1ToL2MessageWriter[] | ClassicL1ToL2MessageReader[]
```

Comment:
@yahgwai @spsjvc to avoid having different data type returned from the function we could either:

- pass version prop to the function ('classic' | 'nitro') and based on that we could return correct types. However this way a user will have to provide the 'classic' flag to get the classic data - they will have to figure out themselves whether the txn is classic or nitro

or:

- structure classic data to have the same keys as nitro, so this way ts won't give warnings/errors when trying to access properties

LMK

Reply:
I think you could do this by extending the L1ToL2MessageReader, and creating a function on the base class for calculating the retryableCreationId and then overriding it in a ClassicL1ToL2MessageReader. Eg.

```
export abstract class L1ToL2Message {

 protected calculateRetryableId(
    chainId: number,
    sender: string,
    messageNumber: BigNumber,
    l1BaseFee: BigNumber,
    messageData: RetryableMessageParams
  ) {
    return L1ToL2Message.calculateSubmitRetryableId(
      chainId,
      sender,
      messageNumber,
      l1BaseFee,
      messageData.destAddress,
      messageData.l2CallValue,
      messageData.l1Value,
      messageData.maxSubmissionFee,
      messageData.excessFeeRefundAddress,
      messageData.callValueRefundAddress,
      messageData.gasLimit,
      messageData.maxFeePerGas,
      messageData.data
    )
  }

  protected constructor(
    public readonly chainId: number,
    public readonly sender: string,
    public readonly messageNumber: BigNumber,
    public readonly l1BaseFee: BigNumber,
    public readonly messageData: RetryableMessageParams
  ) {
    this.retryableCreationId = this.calculateRetryableId(
      chainId,
      sender,
      messageNumber,
      l1BaseFee,
      messageData
    )
  }
}

export class ClassicL1ToL2MessageReader2 extends L1ToL2MessageReader {
  constructor(
    l2Provider: Provider,
    chainId: number,
    sender: string,
    messageNumber: BigNumber,
    l1BaseFee: BigNumber,
    messageData: RetryableMessageParams
  ) {
    super(l2Provider, chainId, sender, messageNumber, l1BaseFee, messageData)
  }

  protected override calculateRetryableId(
    chainId: number,
    sender: string,
    messageNumber: BigNumber,
    l1BaseFee: BigNumber,
    messageData: RetryableMessageParams
  ): string {
    const bitFlip = (num: BigNumber) => num.or(BigNumber.from(1).shl(255))
    return keccak256(
      concat([
        zeroPad(BigNumber.from(chainId).toHexString(), 32),
        zeroPad(bitFlip(this.messageNumber).toHexString(), 32),
      ])
    )
  }
}

```

For the auto redeem id and l2 transaction hash i think you may be able to override other functions too, but I havent looked into it.